module/zfs/dsl_crypt.c

```diff
@@ -1193,10 +1193,10 @@ dsl_crypto_key_sync(dsl_crypto_key_t *dck, dmu_tx_t *tx)
 {
 	zio_crypt_key_t *key = &dck->dck_key;
 	dsl_wrapping_key_t *wkey = dck->dck_wkey;
-	uint8_t keydata[MASTER_KEY_MAX_LEN];
-	uint8_t hmac_keydata[SHA512_HMAC_KEYLEN];
-	uint8_t iv[WRAPPING_IV_LEN];
-	uint8_t mac[WRAPPING_MAC_LEN];
+	uint8_t keydata[MASTER_KEY_MAX_LEN] = {0};
```

Comment:
It is too early to have strong opinions since this PR is WIP, I must warn you though that `= {0}` generates `memset()`/`rep stos`. This might directly impact the performance, depends on how often this function is called in kernel space and how big the arrays are.

Reply:
I am aware that zeroing regions is usually not free, yes.

The claim, and reason for the PR, was that without fixing a number of existing uses in the codebase, trying to use things like `valgrind` just gets drowned out by a bunch of things referencing these regions.

If people think that's not worthwhile, that's fine, I'll make a note and avoid it in the future.

Reply:
On the contrary, valgrind complains for a reason and memory corruption is one of the biggest plagues in C/C++ landscape. Fixing those errors is important. I just wanted to highlight potential unintended consequences.

Reply:
If this is file is shared by both kernel and userspace you can make zeroing conditional on compiling for the latter.

Are you sure the valgrind warning is a false-positive here?

Reply:
This function is only called when updating the key data, this should be a relatively infrequent operation and the function as implemented is already fairly expensive.  This change wont have any meaningful performance impact.  Why don't we at least more forward with this hunk which should be uncontroversial.

I agree, the `zio_write_compress()` change is likely too expensive and doing something reasonable there is going to be more difficult.

Reply:
So again, if this is a false positive or is otherwise safe to ignore for whatever reason, silencing valgrind may be hidden behind an ifdef (can't come up with a good name for it though). Then there is just some annoyance that you have to rebuild if you want to trace it, but given that other types of tooling would require rebuilding anyway to add instrumentation I don't think it's a big deal.

Perhaps ZFS_MEM_DEBUG?